upstairs/src/test.rs

```diff
@@ -5105,4 +5245,970 @@ mod up_test {
+            .await
+            .unwrap();
+
+        // op 1
+        upstairs.submit_flush(None, None).await.unwrap();
+
+        // op 2
+        upstairs
+            .submit_write(
+                Block::new_512(0),
+                Bytes::from(vec![0x55; 512]),
+                None,
+                false,
+            )
+            .await
+            .unwrap();
+
+        let ds = upstairs.downstairs.lock().await;
+        let keys: Vec<&u64> = ds.active.keys().sorted().collect();
+        let jobs: Vec<&DownstairsIO> =
+            keys.iter().map(|k| ds.active.get(k).unwrap()).collect();
+        assert_eq!(jobs.len(), 3);
+
+        assert!(jobs[0].work.deps().is_empty());
+        assert_eq!(jobs[1].work.deps(), &[jobs[0].ds_id]);
+        assert_eq!(jobs[2].work.deps(), &[jobs[1].ds_id]);
+    }
+
+    #[tokio::test]
+    async fn test_deps_all_writes_depend_on_flushes() {
+        // Test that the following job dependency graph is made:
+        //
+        //          block
+        // op# | 0 1 2 3 4 5 | deps
+        // ----|-------------|-----
+        //   0 | W           |
+        //   1 |   W         |
+        //   2 |     W       |
+        //   3 | FFFFFFFFFFF | 0,1,2
+        //   4 |       W     | 3
+        //   5 |         W   | 3
+        //   6 |           W | 3
+
+        let upstairs = make_upstairs();
+        upstairs.set_active().await.unwrap();
+
+        // ops 0 to 2
+        for i in 0..3 {
+            upstairs
+                .submit_write(
+                    Block::new_512(i),
+                    Bytes::from(vec![0xff; 512]),
+                    None,
+                    false,
+                )
+                .await
+                .unwrap();
+        }
+
+        // op 3
+        upstairs.submit_flush(None, None).await.unwrap();
+
+        // ops 4 to 6
+        for i in 3..6 {
+            upstairs
+                .submit_write(
+                    Block::new_512(i),
+                    Bytes::from(vec![0xff; 512]),
+                    None,
+                    false,
+                )
+                .await
+                .unwrap();
+        }
+
+        let ds = upstairs.downstairs.lock().await;
+        let keys: Vec<&u64> = ds.active.keys().sorted().collect();
+        let jobs: Vec<&DownstairsIO> =
+            keys.iter().map(|k| ds.active.get(k).unwrap()).collect();
+        assert_eq!(jobs.len(), 7);
+
+        assert!(jobs[0].work.deps().is_empty()); // write @ 0
+        assert!(jobs[1].work.deps().is_empty()); // write @ 1
+        assert!(jobs[2].work.deps().is_empty()); // write @ 2
+
+        assert_eq!(
+            hashset(jobs[3].work.deps()), // flush
+            hashset(&[jobs[0].ds_id, jobs[1].ds_id, jobs[2].ds_id]),
+        );
+
+        assert_eq!(jobs[4].work.deps(), &[jobs[3].ds_id]); // write @ 3
+        assert_eq!(jobs[5].work.deps(), &[jobs[3].ds_id]); // write @ 4
+        assert_eq!(jobs[6].work.deps(), &[jobs[3].ds_id]); // write @ 5
+    }
+
```

Comment:
What about mixing some reads and writes, like these:
```
//       block
// op# | 0 1 2 3 4 5 | deps
// ----|-------------|-----
//   0 | R           |
//   1 | W           |

```
Or this more complicated one:
```
//       block
// op# | 0 1 2 3 4 5 | deps
// ----|-------------|-----
//   0 | R           |
//   1 | W           |
//   2 |   R R       |
//   3 |   W W       |
//   4 |       R R   |
//   5 |       W W   |
```

Reply:
done in c9faeeb